### PR TITLE
Deepspeed integration.

### PR DESCRIPTION
**Pull request by hyunwoongko**
Please see `MODEL_SCALABILITY.md` to check detailed information of this PR ;)

### CODE REVIEW

File: MODEL_SCALABILITY.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+       --deepspeed_config ./configs/scalability/YOUR_DS_CONFIG_NAME.json
+```
+
+## 3. Integrated architectures
+- `carp.py`
+- `carp_cloob.py`
+- `carp_coop.py`
+- `carp_shared_encoder.py`
+
+`carp_mlm.py` and `carp_momentum.py` are not integrated yet.
```
Comment (LouisCastricato) [inline MODEL_SCALABILITY.md:33]: We don't need to mention these, these are both failed experiments.

File: carp/pytorch/model/architectures/__init__.py [hunk 2]
```diff
@@ -202,71 +208,100 @@ def encode_passages(self, x):
...
-            
-    def step(self,
-        scaler : torch.cuda.amp.GradScaler,
-        opt: torch.optim.Optimizer):
-        if self.accum_step % self.config.grad_accum == 0:
-            scaler.step(opt)
-            scaler.update()
-            self.accum_step = 0
+
+class BaseTrainer(object):
```
Comment (LouisCastricato) [inline carp/pytorch/model/architectures/__init__.py]: I really like the trainer abstraction! Good idea :)

File: carp/pytorch/model/architectures/carp_cloob.py [hunk 3]
```diff
@@ -99,47 +120,115 @@ def train_step(
...
+                    review.to(self.model.module.device),
+                    mask.to(self.model.module.device),
+                )  # grad _just_ at positions in `index`
+            loss = self.model.module.cloob(
+                torch.cat(forward_output["pass_encs"]), torch.cat(rev_tmp)
+            )
+            self.model.backward(loss)
+
+        self.model.step()
+        # Kevin: Why not accumulation?
```
Comment (LouisCastricato) [inline carp/pytorch/model/architectures/carp_cloob.py:174]: Oops, I forgot to add it here. Thanks.

**Comment by LouisCastricato** [comment]
Looks solid, I am just not 100% convinced that a trainer is the best abstraction. Do you think we should combine trainer with orchestrator?

File: MODEL_SCALABILITY.md (hunk 1 above)
Comment (hyunwoongko) [reply MODEL_SCALABILITY.md:33]: I see

**Comment by hyunwoongko** [comment]
> Do you think we should combine trainer with orchestrator?

it's good idea. let's discuss more.

File: MODEL_SCALABILITY.md (hunk 1 above)
Comment (hyunwoongko) [reply MODEL_SCALABILITY.md:33]: let's remove or move to legacy folder.
please lmk which experiments were failed (no more used)

File: carp/pytorch/model/architectures/carp_cloob.py (hunk 3 above)
Comment (LouisCastricato) [reply carp/pytorch/model/architectures/carp_cloob.py:174]: Resolved

File: MODEL_SCALABILITY.md (hunk 1 above)
Comment (LouisCastricato) [reply MODEL_SCALABILITY.md:33]: Going to do this in a future PR
